interface/main/main_screen.php

```diff
@@ -28,149 +28,265 @@ function posted_to_hidden($name)
+            // We have key data, check if it matches what was registered.
+            $tmprow = sqlQuery("SELECT login_work_area FROM users_secure WHERE id = ?", array($userid));
+            try {
+                $registration = $u2f->doAuthenticate(
+                    json_decode($tmprow['login_work_area']), // these are the original challenge requests
+                    $registrations,
+                    json_decode($_POST['form_response'])
+                );
+                // Stored registration data needs to be updated because the usage count has changed.
+                // We have to use the matching registered key.
+                $strhandle = json_encode($registration->keyHandle);
+                if (isset($regs[$strhandle])) {
+                    sqlStatement(
+                        "UPDATE login_mfa_registrations SET `var1` = ? WHERE " .
+                        "`user_id` = ? AND `method` = 'U2F' AND `name` = ?",
+                        array(json_encode($registration), $userid, $regs[$strhandle])
+                    );
+                } else {
+                    error_log("Unexpected keyHandle returned from doAuthenticate(): '$strhandle'");
+                }
+                // Keep track of when challenges were last answered correctly.
+                sqlStatement(
+                    "UPDATE users_secure SET last_challenge_response = NOW() WHERE id = ?",
+                    array($_SESSION['authId'])
+                );
+            } catch (u2flib_server\Error $e) {
+                // Authentication failed so we will build the U2F form again.
+                $form_response = '';
+                $errormsg = xl('U2F Key Authentication error') . ": " . $e->getMessage();
+                $errortype = "UTF";
+            }
+        } else {
+            // do nothing
             $form_response = '';
-            $errormsg = xl('Authentication error') . ": " . $e->getMessage();
         }
     }
+
     if (!$form_response) {
-        // There is no key data yet or authentication failed, so we need to solicit it.
-        $requests = json_encode($u2f->getAuthenticateData($registrations));
-        // Persist the challenge also in the database because the browser is untrusted.
-        sqlStatement(
-            "UPDATE users_secure SET login_work_area = ? WHERE id = ?",
-            array($requests, $userid)
-        );
-        generate_html_start(xl('U2F Key Verification'));
-        echo "<p>\n";
-        echo xlt('Insert your key into a USB port and click the Authenticate button below.');
-        echo " " . xlt('Then press the flashing button on your key within 1 minute.') . "</p>\n";
-        echo "<table><tr><td>\n";
-        echo "<input type='button' value='" . xla('Authenticate') . "' onclick='doAuth()' />\n";
-        echo "<input type='hidden' name='form_requests' value='" . attr($requests) . "' />\n";
-        echo "<input type='hidden' name='form_response' value='' />\n";
```

Comment:
The form_response input field is now missing in the U2F-only case, so authentication fails.

Reply:
Yep there's a javascript error when the U2F device button is pressed. function doAuth() fails because f.form_response does not exist.